samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,284 @@
+    AWS_SERVERLESS_STATEMACHINE,
+    AWS_STEPFUNCTIONS_STATEMACHINE,
+    SYNCABLE_RESOURCES,
+)
+from samcli.yamlhelper import yaml_parse
+
+LOG = logging.getLogger(__name__)
+
+GENERAL_REMOVAL_MAP = {
+    AWS_SERVERLESS_FUNCTION: ["CodeUri", "ImageUri"],
+    AWS_SERVERLESS_LAYERVERSION: ["ContentUri"],
+    AWS_LAMBDA_LAYERVERSION: ["Content"],
+    AWS_SERVERLESS_API: ["DefinitionBody"],
+    AWS_APIGATEWAY_RESTAPI: ["BodyS3Location"],
+    AWS_SERVERLESS_HTTPAPI: ["DefinitionUri"],
+    AWS_APIGATEWAY_V2_API: ["BodyS3Location"],
+    AWS_SERVERLESS_STATEMACHINE: ["DefinitionUri"],
+    AWS_STEPFUNCTIONS_STATEMACHINE: ["DefinitionS3Location"],
+}
+
+LAMBDA_FUNCTION_REMOVAL_MAP = {
+    AWS_LAMBDA_FUNCTION: {"Code": ["ImageUri", "S3Bucket", "S3Key", "S3ObjectVersion"]},
+}
+
+
+class InfraSyncExecutor:
+    """
+    Executor for infra sync that contains skip logic when template is not changed
+    """
+
+    _build_context: BuildContext
+    _package_context: PackageContext
+    _deploy_context: DeployContext
+
+    def __init__(self, build_context: BuildContext, package_context: PackageContext, deploy_context: DeployContext):
+        """Constructs the sync for infra executor.
+
+        Parameters
+        ----------
+        build_context : BuildContext
+            BuildContext
+        package_context : PackageContext
+            PackageContext
+        deploy_context : DeployContext
+            DeployContext
+        """
+        self._build_context = build_context
+        self._package_context = package_context
+        self._deploy_context = deploy_context
+
+        self._session = Session(profile_name=self._deploy_context.profile, region_name=self._deploy_context.region)
+        self._cfn_client = self._boto_client("cloudformation")
+        self._s3_client = self._boto_client("s3")
+
+    def _boto_client(self, client_name: str):
+        """
+        Creates boto client
+
+        Parameters
+        ----------
+        client_name: str
+            The name of the client
+
+        Returns
+        -------
+        Service client instance
+        """
+        return get_boto_client_provider_from_session_with_config(self._session)(client_name)
+
+    def _compare_templates(self, local_template_path: str, stack_name: str) -> bool:
+        """
+        Recursively conpares two templates, including the nested templates referenced inside
+
+        Parameters
+        ----------
+        local_template_path : str
+            The local template location
+        stack_name : str
+            The CloudFormation stack name that the template is deployed to
+
+        Returns
+        -------
+        bool
+            Returns True if two templates are identical
+            Returns False if two templates are different
+        """
+        current_template = None
+        # If the customer template uses a nested stack with location/template URL in S3
+        if local_template_path.startswith("https://"):
```

Comment:
Would there be a case where template will start with `s3://` notation?

Reply:
I'll take a look at this field, if notation with s3:// is supported, I will add parsing logic here as well

Reply:
With experimentation, we found that there are three formats that the Location field can take: 1. local template path, which we will package in package step. 2. s3 URL, which is a https:// location. 3. ApplicationLocationObject, which is a dict that contains SAR application ID. I have added new logic that handles the 3rd scenario, now all possible values are covered